signalling/src/server/mod.rs

```diff
@@ -0,0 +1,227 @@
+        if let Some(peer) = self.peers.get_mut(peer_id) {
+            let mut sender = peer.sender.clone();
+            let peer_id = peer_id.to_string();
+            task::spawn(async move {
+                if let Err(err) = sender.send(message).await {
+                    warn!(peer_id = %peer_id, "Failed to send message: {}", err);
+                }
+            });
+        }
+    }
+}
+
+struct State {
+    message_handler: Box<dyn MessageHandler>,
+    message_sender: Option<DefaultMessageSender>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum SignallingServerError {}
+
+impl SignallingServer {
+    /// Instantiate the signalling server with a MessageHandler
+    /// implementation. Use DefaultMessageHandler for the default
+    /// protocol, or implement your own handler
+    pub fn new(message_handler: Box<dyn MessageHandler>) -> Self {
+        Self {
+            state: Arc::new(Mutex::new(State {
+                message_handler,
+                message_sender: Some(DefaultMessageSender::default()),
+            })),
+        }
+    }
+
+    fn remove_peer(state: Arc<Mutex<State>>, peer_id: &str) {
+        {
+            let mut state = state.lock().unwrap();
+            let mut message_sender = state.message_sender.take().unwrap();
+
+            state
+                .message_handler
+                .remove_peer(&mut message_sender, peer_id);
+
+            state.message_sender = Some(message_sender);
+        }
+
+        if let Some(mut peer) = state
+            .lock()
+            .unwrap()
+            .message_sender
+            .as_mut()
+            .unwrap()
+            .peers
+            .remove(peer_id)
+        {
+            let peer_id = peer_id.to_string();
+            task::spawn(async move {
+                peer.sender.close_channel();
+                if let Err(err) = peer.send_task_handle.await {
+                    trace!(peer_id = %peer_id, "Error while joining send task: {}", err);
+                }
+                peer.receive_task_handle.await;
+            });
+        }
+    }
+
+    /// Can be used to inject messages constructed locally
+    pub fn handle_message(&self, msg: String, peer_id: &str) {
+        let mut state = self.state.lock().unwrap();
+        let mut message_sender = state.message_sender.take().unwrap();
+
+        if let Err(err) = state
+            .message_handler
+            .handle_message(&mut message_sender, msg, peer_id)
+        {
+            warn!(this = %peer_id, "Error handling message: {:?}", err);
+            message_sender.send_message(
+                serde_json::to_string(&p::OutgoingMessage::Error {
+                    details: err.to_string(),
+                })
+                .unwrap(),
+                peer_id,
+            );
+        }
+
+        state.message_sender = Some(message_sender);
+    }
+
+    async fn accept_connection(state: Arc<Mutex<State>>, stream: TcpStream) {
```

Comment:
Should this maybe take an `AsyncRead + AsyncWrite` to be more generic? Otherwise you need to change things here depending on whether TLS is used or not